File overwrite error with Exasol 6.2.14

## Situation

When running a matrix test against Exasol 6.2.14 the sync check fails with the following error:
```
2021-03-29 14:05:03.083 [WARNING] Leaving container running since reuse is enabled. Don't forget to stop and remove the container manually using docker rm -f CONTAINER_ID. 
[ERROR] Tests run: 14, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 218.208 s <<< FAILURE! - in com.exasol.bucketfs.SyncAwareBucketIT
[ERROR] testReplaceFile{Path}  Time elapsed: 5.61 s  <<< FAILURE!
java.lang.AssertionError: 
Upload number 2: file B
Expected: "abcdeABCDE\n"
     but: was "0123456789\n"
	at com.exasol.bucketfs.SyncAwareBucketIT.testReplaceFile(SyncAwareBucketIT.java:156)
```

## Acceptance Criteria

* Overwrite test green with Exasol 6.2.14

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,7 +4,8 @@ jdk:
   - openjdk11
 
 env:
-  - DOCKER_DB_VERSION=7.0.6
+  - DOCKER_DB_VERSION=6.2.14
+  - DOCKER_DB_VERSION=7.0.9
 
 addons:
   sonarcloud:

--- a/README.md
+++ b/README.md
@@ -42,41 +42,4 @@ Requirement, design documents and coverage tags are written in [OpenFastTrace](h
 
 * [System Requirement Specification](doc/system_requirements.md)
 * [Design](doc/design.md)
-
-## Dependencies
-
-### Run Time Dependencies
-
-Running the BucketFS Java requires a Java Runtime version 11 or later.
-
-### Test Dependencies
-
-| Dependency                                                                          | Purpose                                                | License                       |
-|-------------------------------------------------------------------------------------|--------------------------------------------------------|-------------------------------|
-| [Java Hamcrest](http://hamcrest.org/JavaHamcrest/)                                  | Checking for conditions in code via matchers           | BSD License                   |
-| [JUnit](https://junit.org/junit5)                                                   | Unit testing framework                                 | Eclipse Public License 1.0    |
-| [Mockito](http://site.mockito.org/)                                                 | Mocking framework                                      | MIT License                   |
-| [Testcontainers](https://www.testcontainers.org/)                                   | Docker Container control abstraction                   | MIT License                   |
-
-### Build Dependencies
-
-| Plug-in                                                                             | Purpose                                                | License                       |
-|-------------------------------------------------------------------------------------|--------------------------------------------------------|-------------------------------|
-| [Apache Maven](https://maven.apache.org/)                                           | Build tool                                             | Apache License 2.0            |
-| [Maven Compiler Plugin](https://maven.apache.org/plugins/maven-compiler-plugin/)    | Setting required Java version                          | Apache License 2.0            |
-| [Maven GPG Plugin](https://maven.apache.org/plugins/maven-gpg-plugin/)              | Code signing                                           | Apache License 2.0            |
-| [Maven Enforcer Plugin][maven-enforcer-plugin]                                      | Controlling environment constants                      | Apache License 2.0            |
-| [Maven Javadoc Plugin](https://maven.apache.org/plugins/maven-javadoc-plugin/)      | Creating a Javadoc JAR                                 | Apache License 2.0            |
-| [Maven Jacoco Plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html)          | Code coverage metering                                 | Eclipse Public License 2.0    |
-| [Maven Source Plugin](https://maven.apache.org/plugins/maven-source-plugin/)        | Creating a source code JAR                             | Apache License 2.0            |
-| [Maven Surefire Plugin](https://maven.apache.org/surefire/maven-surefire-plugin/)   | Unit testing                                           | Apache License 2.0            |
-| [OpenFastTrace Maven Plugin][oft-maven-plugin]                                      | Requirement Tracing                                    | GPL v3                        |
-| [Project Keeper Maven Plugin][project-keeper-maven-plugin]                          | Checking project structure                             | MIT License                   |
-| [Sonatype OSS Index Maven Plugin][sonatype-oss-index-maven-plugin]                  | Checking Dependencies Vulnerability                    | ASL2                          |
-| [Versions Maven Plugin][versions-maven-plugin]                                      | Checking if dependencies updates are available         | Apache License 2.0            |
-
-[maven-enforcer-plugin]: http://maven.apache.org/enforcer/maven-enforcer-plugin/
-[oft-maven-plugin]: https://github.com/itsallcode/openfasttrace-maven-plugin
-[project-keeper-maven-plugin]: https://github.com/exasol/project-keeper-maven-plugin
-[sonatype-oss-index-maven-plugin]: https://sonatype.github.io/ossindex-maven/maven-plugin/
-[versions-maven-plugin]: https://www.mojohaus.org/versions-maven-plugin/
+* [Dependencies](dependencies.md)

--- a/dependencies.md
+++ b/dependencies.md
@@ -1,0 +1,83 @@
+<!-- @formatter:off -->
+# Dependencies
+
+## Compile Dependencies
+
+| Dependency                | License  |
+| ------------------------- | -------- |
+| [error-reporting-java][0] | [MIT][1] |
+
+## Test Dependencies
+
+| Dependency                                      | License                          |
+| ----------------------------------------------- | -------------------------------- |
+| [JUnit Jupiter Engine][2]                       | [Eclipse Public License v2.0][3] |
+| [JUnit Jupiter Params][2]                       | [Eclipse Public License v2.0][3] |
+| [Hamcrest][6]                                   | [BSD License 3][7]               |
+| [mockito-junit-jupiter][8]                      | [The MIT License][9]             |
+| [Test containers for Exasol on Docker][10]      | [MIT][1]                         |
+| [Testcontainers :: JUnit Jupiter Extension][12] | [MIT][13]                        |
+
+## Plugin Dependencies
+
+| Dependency                                              | License                                        |
+| ------------------------------------------------------- | ---------------------------------------------- |
+| [Maven Surefire Plugin][14]                             | [Apache License, Version 2.0][15]              |
+| [Maven Failsafe Plugin][16]                             | [Apache License, Version 2.0][15]              |
+| [JaCoCo :: Maven Plugin][18]                            | [Eclipse Public License 2.0][19]               |
+| [Apache Maven Compiler Plugin][20]                      | [Apache License, Version 2.0][15]              |
+| [OpenFastTrace Maven Plugin][22]                        | [GNU General Public License v3.0][23]          |
+| [Project keeper maven plugin][24]                       | [MIT][1]                                       |
+| [Versions Maven Plugin][26]                             | [Apache License, Version 2.0][15]              |
+| [org.sonatype.ossindex.maven:ossindex-maven-plugin][28] | [ASL2][29]                                     |
+| [Apache Maven Enforcer Plugin][30]                      | [Apache License, Version 2.0][15]              |
+| [error-code-crawler-maven-plugin][32]                   | [MIT][1]                                       |
+| [Reproducible Build Maven Plugin][34]                   | [Apache 2.0][29]                               |
+| [Apache Maven GPG Plugin][36]                           | [Apache License, Version 2.0][29]              |
+| [Apache Maven Deploy Plugin][38]                        | [Apache License, Version 2.0][15]              |
+| [Nexus Staging Maven Plugin][40]                        | [Eclipse Public License][41]                   |
+| [Apache Maven Source Plugin][42]                        | [Apache License, Version 2.0][15]              |
+| [Apache Maven Javadoc Plugin][44]                       | [Apache License, Version 2.0][15]              |
+| [Maven Clean Plugin][46]                                | [The Apache Software License, Version 2.0][29] |
+| [Maven Resources Plugin][48]                            | [The Apache Software License, Version 2.0][29] |
+| [Maven JAR Plugin][50]                                  | [The Apache Software License, Version 2.0][29] |
+| [Maven Install Plugin][52]                              | [The Apache Software License, Version 2.0][29] |
+| [Maven Site Plugin 3][54]                               | [The Apache Software License, Version 2.0][29] |
+
+[18]: https://www.eclemma.org/jacoco/index.html
+[24]: https://github.com/exasol/project-keeper-maven-plugin
+[0]: https://github.com/exasol/error-reporting-java
+[29]: http://www.apache.org/licenses/LICENSE-2.0.txt
+[14]: https://maven.apache.org/surefire/maven-surefire-plugin/
+[40]: http://www.sonatype.com/public-parent/nexus-maven-plugins/nexus-staging/nexus-staging-maven-plugin/
+[46]: http://maven.apache.org/plugins/maven-clean-plugin/
+[1]: https://opensource.org/licenses/MIT
+[8]: https://github.com/mockito/mockito
+[16]: https://maven.apache.org/surefire/maven-failsafe-plugin/
+[26]: http://www.mojohaus.org/versions-maven-plugin/
+[7]: http://opensource.org/licenses/BSD-3-Clause
+[20]: https://maven.apache.org/plugins/maven-compiler-plugin/
+[13]: http://opensource.org/licenses/MIT
+[36]: http://maven.apache.org/plugins/maven-gpg-plugin/
+[22]: https://github.com/itsallcode/openfasttrace-maven-plugin
+[19]: https://www.eclipse.org/legal/epl-2.0/
+[38]: https://maven.apache.org/plugins/maven-deploy-plugin/
+[41]: http://www.eclipse.org/legal/epl-v10.html
+[10]: https://github.com/exasol/exasol-testcontainers
+[34]: http://zlika.github.io/reproducible-build-maven-plugin
+[23]: https://www.gnu.org/licenses/gpl-3.0.html
+[50]: http://maven.apache.org/plugins/maven-jar-plugin/
+[15]: https://www.apache.org/licenses/LICENSE-2.0.txt
+[30]: https://maven.apache.org/enforcer/maven-enforcer-plugin/
+[9]: https://github.com/mockito/mockito/blob/release/3.x/LICENSE
+[3]: https://www.eclipse.org/legal/epl-v20.html
+[52]: http://maven.apache.org/plugins/maven-install-plugin/
+[2]: https://junit.org/junit5/
+[28]: https://sonatype.github.io/ossindex-maven/maven-plugin/
+[12]: https://testcontainers.org
+[42]: https://maven.apache.org/plugins/maven-source-plugin/
+[6]: http://hamcrest.org/JavaHamcrest/
+[54]: http://maven.apache.org/plugins/maven-site-plugin/
+[48]: http://maven.apache.org/plugins/maven-resources-plugin/
+[44]: https://maven.apache.org/plugins/maven-javadoc-plugin/
+[32]: https://github.com/exasol/error-code-crawler-maven-plugin

--- a/doc/changes/changelog.md
+++ b/doc/changes/changelog.md
@@ -1,4 +1,5 @@
 # Changes
 
+* [2.0.1](changes_2.0.1.md)
 * [2.0.0](changes_2.0.0.md)
 * [1.0.0](changes_1.0.0.md)

--- a/doc/changes/changes_2.0.1.md
+++ b/doc/changes/changes_2.0.1.md
@@ -1,0 +1,20 @@
+# bucketfs-java 2.0.0, released 2021-04-29
+
+Code name: Improved repeated upload reliability
+
+## Summary
+
+In certain situations especially on 6.2.x there was a race condition that made detection of successful reupload to the same path in a bucket unreliable.
+This has now been fixed. This scenario is realistically only happening in integration tests, but there it is likely.
+
+Replaced handwritten dependency list with auto-generated list and linked it in the [README](../../README.md).
+
+## Bugfixes
+
+* #4: Fixed race condition on repeated fast upload to the same path in a bucket.
+
+## Dependency Updates
+
+### Plugin Dependency Updates
+
+* Updated `com.exasol:project-keeper-maven-plugin:0.6.1` to `0.7.0`

--- a/doc/changes/changes_2.0.1.md
+++ b/doc/changes/changes_2.0.1.md
@@ -1,4 +1,4 @@
-# bucketfs-java 2.0.1, released 2021-04-??
+# bucketfs-java 2.0.1, released 2021-04-30
 
 Code name: Improved repeated upload reliability
 

--- a/doc/changes/changes_2.0.1.md
+++ b/doc/changes/changes_2.0.1.md
@@ -7,7 +7,7 @@ Code name: Improved repeated upload reliability
 In certain situations especially on 6.2.x there was a race condition that made detection of successful reupload to the same path in a bucket unreliable.
 This has now been fixed. This scenario is realistically only happening in integration tests, but there it is likely.
 
-Replaced handwritten dependency list with auto-generated list and linked it in the [README](../../README.md).
+Replaced handwritten dependency list with auto-generated list and linked it in the README.
 
 ## Bugfixes
 

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
     <modelVersion>4.0.0</modelVersion>
     <groupId>com.exasol</groupId>
     <artifactId>bucketfs-java</artifactId>
-    <version>2.0.0</version>
+    <version>2.0.1</version>
     <name>BucketFS Java</name>
     <description>Java library for automating tasks on Exasol's BucketFS.
     </description>
@@ -224,7 +224,7 @@
             <plugin>
                 <groupId>com.exasol</groupId>
                 <artifactId>project-keeper-maven-plugin</artifactId>
-                <version>0.6.1</version>
+                <version>0.7.0</version>
                 <executions>
                     <execution>
                         <goals>

--- a/src/main/java/com/exasol/bucketfs/SyncAwareBucket.java
+++ b/src/main/java/com/exasol/bucketfs/SyncAwareBucket.java
@@ -40,6 +40,7 @@ public class SyncAwareBucket extends WriteEnabledBucket implements Bucket {
         final var millisSinceEpochBeforeUpload = System.currentTimeMillis();
         uploadFileNonBlocking(localPath, pathInBucket);
         waitForFileToBeSynchronized(pathInBucket, millisSinceEpochBeforeUpload);
+        recordUploadInHistory(pathInBucket);
     }
 
     // Wait some time between uploads of the same file so we can distinguish the upload success logs for detecting
@@ -50,7 +51,10 @@ public class SyncAwareBucket extends WriteEnabledBucket implements Bucket {
             final Instant lastUploadAt = this.uploadHistory.get(extendedPathInBucket) //
                     .with(ChronoField.NANO_OF_SECOND, 0);
             final Instant now = Instant.now();
-            if (!now.isAfter(lastUploadAt.plusSeconds(1))) {
+            if (now.isAfter(lastUploadAt.plusSeconds(1))) {
+                LOGGER.fine(() -> "Last upload to '" + extendedPathInBucket + "' was at " + lastUploadAt
+                        + ". No need to add extra delay.");
+            } else {
                 final long delayInMillis = 1000L - (now.getNano() / 1000000L);
                 LOGGER.fine(() -> "Delaying upload to '" + extendedPathInBucket + "' for " + delayInMillis + " ms");
                 try {
@@ -62,7 +66,9 @@ public class SyncAwareBucket extends WriteEnabledBucket implements Bucket {
                             .toString());
                 }
             }
-        } else {
+        } else
+
+        {
             LOGGER.fine(() -> "No previous uploads to '" + extendedPathInBucket
                     + "' recorded in upload history. No upload delay required.");
         }
@@ -76,6 +82,7 @@ public class SyncAwareBucket extends WriteEnabledBucket implements Bucket {
         final var millisSinceEpochBeforeUpload = System.currentTimeMillis();
         uploadStringContentNonBlocking(content, pathInBucket);
         waitForFileToBeSynchronized(pathInBucket, millisSinceEpochBeforeUpload);
+        recordUploadInHistory(pathInBucket);
     }
 
     // [impl->dsn~uploading-input-stream-to-bucket~1]
@@ -86,6 +93,7 @@ public class SyncAwareBucket extends WriteEnabledBucket implements Bucket {
         final var millisSinceEpochBeforeUpload = System.currentTimeMillis();
         uploadInputStreamNonBlocking(inputStreamSupplier, pathInBucket);
         waitForFileToBeSynchronized(pathInBucket, millisSinceEpochBeforeUpload);
+        recordUploadInHistory(pathInBucket);
     }
 
     // [impl->dsn~waiting-until-archive-extracted~1]

--- a/src/main/java/com/exasol/bucketfs/SyncAwareBucket.java
+++ b/src/main/java/com/exasol/bucketfs/SyncAwareBucket.java
@@ -66,9 +66,7 @@ public class SyncAwareBucket extends WriteEnabledBucket implements Bucket {
                             .toString());
                 }
             }
-        } else
-
-        {
+        } else {
             LOGGER.fine(() -> "No previous uploads to '" + extendedPathInBucket
                     + "' recorded in upload history. No upload delay required.");
         }

--- a/src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java
+++ b/src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java
@@ -40,13 +40,12 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
             throws BucketAccessException, TimeoutException, FileNotFoundException {
         final var extendedPathInBucket = extendPathInBucketDownToFilename(localPath, pathInBucket);
         final var uri = createWriteUri(extendedPathInBucket);
-        uploadWithBodyPublisher(extendedPathInBucket, uri, BodyPublishers.ofFile(localPath),
-                "file '" + localPath + "'");
+        uploadWithBodyPublisher(uri, BodyPublishers.ofFile(localPath), "file '" + localPath + "'");
         recordUploadInHistory(pathInBucket);
     }
 
-    protected void uploadWithBodyPublisher(final String pathInBucket, final URI uri, final BodyPublisher publisher,
-            final String what) throws BucketAccessException {
+    protected void uploadWithBodyPublisher(final URI uri, final BodyPublisher publisher, final String what)
+            throws BucketAccessException {
         LOGGER.fine(() -> "Uploading " + what + " to bucket '" + this + "' at '" + uri + "'");
         requestUpload(uri, publisher);
         LOGGER.fine(() -> "Successfully uploaded " + what + " to '" + uri + "'");
@@ -88,6 +87,15 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
                 messageBuilder("E-BFSJ-6").message("Interrupted trying to upload {{URI}}.", uri).toString());
     }
 
+    /**
+     * Record the upload in the internal upload history of this bucket object.
+     * <p>
+     * This is mainly necessary to work around issues with detecting repeated uploads. Under certain circumstances they
+     * need to be delayed and the upload history helps with that.
+     * </p>
+     *
+     * @param pathInBucket path in the bucket to which an upload happened.
+     */
     protected void recordUploadInHistory(final String pathInBucket) {
         final var now = Instant.now();
         LOGGER.fine(() -> "Recorded upload to '" + pathInBucket + "' at " + now + " in upload history");
@@ -104,8 +112,7 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
             throws BucketAccessException, TimeoutException {
         final var uri = createWriteUri(pathInBucket);
         final var excerpt = (content.length() > 20) ? content.substring(0, 20) + "..." : content;
-        uploadWithBodyPublisher(pathInBucket, uri, BodyPublishers.ofString(content),
-                "string content '" + excerpt + "'");
+        uploadWithBodyPublisher(uri, BodyPublishers.ofString(content), "string content '" + excerpt + "'");
         recordUploadInHistory(pathInBucket);
     }
 
@@ -114,8 +121,7 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
     public void uploadInputStreamNonBlocking(final Supplier<InputStream> inputStreamSupplier, final String pathInBucket)
             throws BucketAccessException, TimeoutException {
         final var uri = createWriteUri(pathInBucket);
-        uploadWithBodyPublisher(pathInBucket, uri, BodyPublishers.ofInputStream(inputStreamSupplier),
-                "content of input stream");
+        uploadWithBodyPublisher(uri, BodyPublishers.ofInputStream(inputStreamSupplier), "content of input stream");
         recordUploadInHistory(pathInBucket);
     }
 

--- a/src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java
+++ b/src/main/java/com/exasol/bucketfs/WriteEnabledBucket.java
@@ -42,13 +42,13 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
         final var uri = createWriteUri(extendedPathInBucket);
         uploadWithBodyPublisher(extendedPathInBucket, uri, BodyPublishers.ofFile(localPath),
                 "file '" + localPath + "'");
+        recordUploadInHistory(pathInBucket);
     }
 
     protected void uploadWithBodyPublisher(final String pathInBucket, final URI uri, final BodyPublisher publisher,
             final String what) throws BucketAccessException {
         LOGGER.fine(() -> "Uploading " + what + " to bucket '" + this + "' at '" + uri + "'");
         requestUpload(uri, publisher);
-        recordUploadInHistory(pathInBucket);
         LOGGER.fine(() -> "Successfully uploaded " + what + " to '" + uri + "'");
     }
 
@@ -88,7 +88,7 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
                 messageBuilder("E-BFSJ-6").message("Interrupted trying to upload {{URI}}.", uri).toString());
     }
 
-    private void recordUploadInHistory(final String pathInBucket) {
+    protected void recordUploadInHistory(final String pathInBucket) {
         final var now = Instant.now();
         LOGGER.fine(() -> "Recorded upload to '" + pathInBucket + "' at " + now + " in upload history");
         this.uploadHistory.put(pathInBucket, now);
@@ -106,6 +106,7 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
         final var excerpt = (content.length() > 20) ? content.substring(0, 20) + "..." : content;
         uploadWithBodyPublisher(pathInBucket, uri, BodyPublishers.ofString(content),
                 "string content '" + excerpt + "'");
+        recordUploadInHistory(pathInBucket);
     }
 
     // [impl->dsn~uploading-input-stream-to-bucket~1]
@@ -115,6 +116,7 @@ public class WriteEnabledBucket extends ReadEnabledBucket implements Unsynchroni
         final var uri = createWriteUri(pathInBucket);
         uploadWithBodyPublisher(pathInBucket, uri, BodyPublishers.ofInputStream(inputStreamSupplier),
                 "content of input stream");
+        recordUploadInHistory(pathInBucket);
     }
 
     /**

--- a/src/test/java/com/exasol/config/BucketFsServiceConfigurationTest.java
+++ b/src/test/java/com/exasol/config/BucketFsServiceConfigurationTest.java
@@ -44,7 +44,8 @@ class BucketFsServiceConfigurationTest {
 
     @Test
     void testGetBucketConfigurationThrowsExceptionIfBucketDoesNotExist() {
+        final BucketFsServiceConfiguration serviceConfigurationBuilder = BucketFsServiceConfiguration.builder().build();
         assertThrows(IllegalArgumentException.class,
-                () -> BucketFsServiceConfiguration.builder().build().getBucketConfiguration("non-existent"));
+                () -> serviceConfigurationBuilder.getBucketConfiguration("non-existent"));
     }
 }

